pycc/local.py

```diff
@@ -355,23 +362,420 @@ def _build_LPNO(self):
+        '''
+         Constructing the approximated perturbed pair density
+        
+        Notes
+        -----
+        Modification of Ruhee's construction of the pno++ density 
+        For now, not frequency dependent (No +omega in the energy denominators)
+        pert = "mu" only
+
+        '''
+        o = slice(0, self.no)
+        v = slice(self.no, self.no+self.nv)
+        print("Perturbed density on. Preparing perturbed density PNOs.")
+
+        # Hbar_ii  = f_ii + t_inef ( 2 * <in|ef> - <in|fe> )
+        Hbar_oo = self.H.F[o,o].copy()
+        Hbar_oo += contract ('inef,mnef->mi',t2, self.H.L[o,o,v,v])
+        Hbar_ii = Hbar_oo.diagonal().copy() 
+       
+        # Hbar_aa = f_aa - t_mnfa (2 * <mn|fa> - <mn|af> )
+        Hbar_vv = self.H.F[v,v].copy()
+        Hbar_vv -= contract ('mnfa,mnfe->ae',t2,self.H.L[o,o,v,v])
+        Hbar_aa = Hbar_vv.diagonal().copy()
+        denom_ia = Hbar_ii.reshape(-1,1) - Hbar_aa
+
+        #need to add this for response
+        #denom_ia += omega
+                
+        denom_ijab = Hbar_ii.reshape(-1, 1, 1, 1) + Hbar_ii.reshape(-1, 1, 1) - Hbar_aa.reshape(-1, 1) - Hbar_aa
+
+        #going to ignore the omega for a moment such that it is a static case
+        #denom_ijab += omega
+
+        self.denom_tuple = (denom_ia, denom_ijab)
+
+        # need to initialize the Abar terms
+        A_list = {}
+        X_guess = {}
+        D = np.zeros((self.no * self.no, self.nv, self.nv))
+        Q_full = np.zeros_like(D)
+        Q = []
+        occ = np.zeros((self.no*self.no, self.nv))
+        dim = np.zeros((self.no*self.no), dtype=int)  # dimension of LPNO space for each pair
+        L = []
+        eps = []
+        T2_full = 0
+        T2_PNO = 0
+        ## Here, perturbation is dipole moment
+        
+        pert = "mu"    
+        if pert == "mu":
+            bs = self.H.basisset
+            mints = psi4.core.MintsHelper(bs) 
+            dipole_array = mints.ao_dipole()
+            dirn = ['X','Y','Z']
+            for i, drn in enumerate(dirn):
+                A_list[drn] = np.einsum('uj,vi,uv', self.C, self.C, np.asarray(dipole_array[i]))
+            
+                A = A_list[drn]
+                # Build guess Abar
+                # Abar_ijab = P_ij^ab (t_ij^eb A_ae - t_mj^ab A_mi)
+                Avvoo = contract('ijeb,ae->abij', t2, A[self.no:, self.no:])
+                Avvoo -= contract('mjab,mi->abij', t2, A[:self.no, :self.no])
+                Abar = Avvoo.swapaxes(0,2).swapaxes(1,3)
+                Abar += Abar.swapaxes(0,1).swapaxes(2,3)
+
+                # Build guess X's
+                # X_ijab = Abar_ijab / Hbar_ii + Hbar_jj - Hbar_aa _ Hbar_bb
+                X_guess[drn] = Abar.copy()
+                X_guess[drn] /= denom_ijab
+
+                D += self.form_density(X_guess[drn])
+  
+            D /= 3.0
+        #print('Average density: {}'.format(D))
+        # now obtain the Q and L 
+        for ij in range(self.no*self.no):
```

Comment:
is there a way you can form the density, then push it into a general build-NOs-from-density function? this could reduce duplicated code in `_LPNO_space` and `_LPNOpp_space`.